File: 0051-n-queens/0051-n-queens.cpp
class Solution {
public:
    
    vector<vector<string>>answer;
    
    bool isPossible(vector<vector<char>>&board,int row,int col,vector<vector<int>>&queenCells){
        
        for(int i=0;i<queenCells.size();++i){
            int nrow = queenCells[i][0];
            int ncol = queenCells[i][1];
            // Same row and column check
            if((row == nrow) || (col == ncol)){
                return false;
            }
            // Diagonal Check
            if((row-col) == (nrow-ncol)){
                return false;
            }
            if((row+col) == (nrow+ncol)){
                return false;
            }
        }
        
        return true;
        
    }
    
    void recursion(vector<vector<char>>&board,int row,int n,vector<vector<int>>&queenCells){
        
        if(row == n){
            vector<string>temp;
            for(int i=0;i<n;i++){
                string curr = "";
                for(int j=0;j<n;j++){
                    curr += board[i][j];
                }
                temp.push_back(curr);
            }
            answer.push_back(temp);
            return;
        }
        
        for(int col=0;col<n;col++){
            if(board[row][col] == '.'){
                if(isPossible(board,row,col,queenCells)){
                    board[row][col] = 'Q';
                    queenCells.push_back({row,col});
                    recursion(board,row+1,n,queenCells);
                    queenCells.pop_back();
                    board[row][col] = '.';
                }
            }
        }
        
    }
    
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<int>>queenCells;
        vector<vector<char>>board(n,vector<char>(n,'.'));
        recursion(board,0,n,queenCells);
        return answer;
    }
};

File: 0051-n-queens/0051-n-queens.java
class Solution {
    
    public boolean isSafe(char board[][], int row, int col, int n){
        for(int i=0;i<n;i++){
            if(board[row][i] == 'Q') return false;
        }
        for(int i=0;i<n;i++){
            if(board[i][col] == 'Q') return false;
        }
        int r = row;
        int c = col;
        for(;r>=0 && c<n;r--,c++){
            if(board[r][c] == 'Q') return false;
        }
        r = row;
        c = col;
        for(;r>=0 && c>=0;r--,c--){
            if(board[r][c] == 'Q') return false;
        }
        r = row;
        c = col;
        for(;r<n && c>=0;r++,c--){
            if(board[r][c] == 'Q') return false;
        }
        r = row;
        c = col;
        for(;r<n && c<n;r++,c++){
            if(board[r][c] == 'Q') return false;
        }
        return true;
    }
    
    public void storeBoard(char board[][], List<List<String>>allboards, int n){
        List<String>rowBoard = new ArrayList<>();
        for(int i=0;i<n;i++){
            String row = "";
            for(int j=0;j<n;j++){
                row += board[i][j]; 
            }
            rowBoard.add(row);
        }
        allboards.add(rowBoard);
    }
    
    public void solve(char board[][], List<List<String>>allboards, int row, int n){
        
        if(row == n){
            storeBoard(board, allboards, n);
            return;
        }
        
        for(int col=0;col<n;col++){
            if(board[row][col] == '.'){
                if(isSafe(board,row,col,n) == true){
                    board[row][col] = 'Q';
                    solve(board, allboards, row + 1, n);
                    board[row][col] = '.';
                }
            }
        }
        
    } 
    
    public List<List<String>> solveNQueens(int n) {
        char [][]board = new char[n][n];
        List<List<String>>allboards = new ArrayList<>();
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                board[i][j] = '.';
            }
        }
        solve(board,allboards,0,n);
        return allboards;
    }
}